### PR TITLE
qutebrowser: add module

### PR DESCRIPTION
**Pull request by JustinLovinger**
A simple Qutebrowser module following RFC 42.

### CODE REVIEW

File: modules/programs/qutebrowser.nix [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+in {
+  options.programs.qutebrowser = {
+    enable = lib.mkEnableOption "programs.qutebrowser";
+
+    settings = lib.mkOption {
+      type = lib.types.attrs;
+      default = { };
+      description = ''
+        Options to add to Qutebrowser <filename>config.py</filename> file.
+        See <link xlink:href="qute://help/settings.html">qute://help/settings.html</link> for options.
```
Comment (rycee) [inline modules/programs/qutebrowser.nix:29]: I would suggest using HTTP links here instead to allow people not using qute at the moment to also read the documentation. For example
```suggestion
        See <link xlink:href="https://qutebrowser.org/doc/help/settings.html"/> for options.
```

File: modules/programs/qutebrowser.nix [hunk 2]
```diff
@@ -0,0 +1,61 @@
+{ config, lib, pkgs, ... }:
+let
```
Comment (rycee) [inline modules/programs/qutebrowser.nix:2]: I would suggest adding a `with lib;` here to avoid having to use `lib.` all over.

**Comment by terlar** [comment]
I did have a WIP on this, that I was meaning to submit, when it was done, that is basically finished, I just needed to add support for bindings. I also wanted to add support for the dictionaries being installed if they are configured `spellcheck.languages`.

I am not sure if the more dynamic or static approach is preferred. Because this module gets quite complicated, as it specifies all options explicitly that is available, see:
https://github.com/terlar/nix-config/blob/dd5305be617235d1cc13a8f972f03d9078610d66/modules/home/programs/qutebrowser.nix

What do you think, is it waste of effort, not desired to have that level of specification? Or does it provide value over the more simple dynamic approach presented here?

File: modules/programs/qutebrowser.nix [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+          ''"${v}"''
+        else
+          builtins.toString v;
+    in if builtins.isAttrs v then
+      lib.concatStringsSep "\n" (lib.mapAttrsToList (formatLine "${o}${n}.") v)
+    else
+      "${o}${n} = ${formatValue v}";
+in {
+  options.programs.qutebrowser = {
+    enable = lib.mkEnableOption "programs.qutebrowser";
```
Comment (rycee) [inline modules/programs/qutebrowser.nix:22]:
```suggestion
    enable = lib.mkEnableOption "qutebrowser";
```

**Comment by rycee** [comment]
@terlar Yeah, unfortunately this extensive level of modelling of the configuration format is not recommended. Some background information about this and a description of the suggested alternative (basically to use a `settings` option much like in this PR) see [RFC 42](https://github.com/NixOS/rfcs/pull/42). The main issue is that an extensive list of options is brittle, it easily breaks when a user runs a different version of the program from what the module author uses.

That doesn't prevent you from using your module, you can for example publish it through [NUR](https://github.com/nix-community/NUR/).

File: modules/programs/qutebrowser.nix (hunk 1 above)
Comment (JustinLovinger) [reply modules/programs/qutebrowser.nix]: I didn't know setting are available there :) Done.

File: modules/programs/qutebrowser.nix (hunk 2 above)
Comment (JustinLovinger) [reply modules/programs/qutebrowser.nix:5]: It's not my personal preference, but I understand the importance of a consistent codebase. Done.

**Comment by JustinLovinger** [comment]
@rycee All of the suggested changes have been made.

**Comment by terlar** [comment]
@rycee Thank you for the info, makes sense.

What about generic converting the casing is that also something to avoid?

E.g. apply this to the strings in this case:
```
camelCaseToSnakeCase = replaceStrings upperChars (map (s: "_${s}") lowerChars);
```

@JustinLovinger Does this work with the dict data-structures, such as:
```
c.aliases = {
    'mpv': 'spawn --userscript view_in_mpv',
    'q': 'quit',
    'w': 'session-save',
    'wq': 'quit --save'
}
c.content.javascript.log = {
    'error': 'debug',
    'info': 'debug',
    'unknown': 'none',
    'warning': 'debug'
}
```

**Comment by JustinLovinger** [comment]
@terlar You can add options that aren't supported by `settings` to `extraConfig`. I couldn't think of a way to disambiguate a Nix set from a Python dict. Maybe if a string starts with '{' or '[', it can avoid wrapping the string in quotes. Then you could add dict values like

```
settings.aliases = ''
  {
    'mpv': 'spawn --userscript view_in_mpv',
    'q': 'quit',
    'w': 'session-save',
    'wq': 'quit --save'
  }
'';
```

However, that might cause issues if you need an actual string that starts with '{' or '['. Such design decisions may be better left for a future pull request. You can always use `extraConfig`.

**Comment by rycee** [comment]
Thanks! Rebased to master in b7737f17322e648a7739529c4728b99b4b65ff5b. I took the liberty of making a few cleanups here and there as well as adding a test case.

**Pull request closed without merging**